core/shared.hpp

```diff
@@ -102,118 +60,41 @@
-#define IRESEARCH_API
-#endif  // IRESEARCH_DLL
-
 // MSVC 2015 does not define __cpp_lib_generic_associative_lookup macro
 #if (defined(__cpp_lib_generic_associative_lookup) || \
      (defined(_MSC_VER) && _MSC_VER >= 1900))
 #define IRESEARCH_GENERIC_ASSOCIATIVE_LOOKUP
 #endif
 
-#if defined(__GNUC__)
-#define IRESEARCH_INIT(f)                           \
-  static void f(void) __attribute__((constructor)); \
-  static void f(void)
-#define RESEARCH_FINI(f)                           \
-  static void f(void) __attribute__((destructor)); \
-  static void f(void)
+#ifndef IRS_FUNC_NAME
+#if defined(__clang__) || defined(__GNUC__)
+#define IRS_FUNC_NAME __PRETTY_FUNCTION__
 #elif defined(_MSC_VER)
-#define IRESEARCH_INIT(f)                                       \
-  static void __cdecl f(void);                                  \
-  static int f##_init_wrapper(void) {                           \
-    f();                                                        \
-    return 0;                                                   \
-  }                                                             \
-  __declspec(allocate(".CRT$XCU")) void(__cdecl * f##_)(void) = \
-    f##_init_wrapper;                                           \
-  static void __cdecl f(void)
-#define RESEARCH_FINI(f)                                              \
-  static void __cdecl f(void);                                        \
-  static int f##_fini_wrapper(void) {                                 \
-    atexit(f);                                                        \
-    return 0;                                                         \
-  }                                                                   \
-  __declspec(allocate(".CRT$XCU")) static int(__cdecl * f##_)(void) = \
-    f##_fini_wrapper;                                                 \
-  static void __cdecl f(void)
-#endif
-
-// define function name used for pretty printing
-// NOTE: the alias points to a compile time finction not a preprocessor macro
-#if defined(__FUNCSIG__) || _MSC_FULL_VER >= 193000000
-#define IRESEARCH_CURRENT_FUNCTION __FUNCSIG__
-#elif defined(__PRETTY_FUNCTION__) || defined(__GNUC__) || defined(__clang__)
-#define IRESEARCH_CURRENT_FUNCTION __PRETTY_FUNCTION__
+#define IRS_FUNC_NAME __FUNCSIG__
 #else
-#error "compiler is not supported"
+#define IRS_FUNC_NAME __func__
 #endif
-
-// IRESEARCH_COMPILER_HAS_FEATURE
-#ifndef __has_feature
-#define IRESEARCH_COMPILER_HAS_FEATURE(x) \
-  0  // Compatibility with non-clang compilers.
-#else
-#define IRESEARCH_COMPILER_HAS_FEATURE(x) __has_feature(x)
 #endif
 
-// IRESEARCH_HAS_ATTRIBUTE
-#ifndef __has_attribute
-#define IRESEARCH_HAS_ATTRIBUTE(x) 0
-#else
-#define IRESEARCH_HAS_ATTRIBUTE(x) __has_attribute(x)
-#endif
-
-// IRESEARCH_ATTRIBUTE_NONNULL
-#if IRESEARCH_HAS_ATTRIBUTE(nonnull) || \
-  (defined(__GNUC__) && !defined(__clang__))
+#if __has_attribute(nonnull)
```

Comment:
what if __has_attribute isn't defined?

Reply:
It always defined in supported compilers

Reply:
I don't think it's necessary, but ok